### ISSUE TITLE
Heap out of bounds read in tftp_print()

### ISSUE THREAD

**Issue by fumfel** [the-tcpdump-group/tcpdump#565]
**Heap out of bounds read in tftp_print()**

Tested on Git HEAD:  e2adb166ecaf59555e18787e86ac8c828efa0acc

To reproduce: `tcpdump -r oobr_tftp_print.pcap`

![Crashing test case - please change extension to '.pcap'](https://cloud.githubusercontent.com/assets/5985287/20092000/0e156562-a596-11e6-957a-21fc45ad42b6.png)

ASAN output:

```
==1391==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x61d00001f281 at pc 0x00000068b575 bp 0x7ffdf90e95a0 sp 0x7ffdf90e9598
READ of size 1 at 0x61d00001f281 thread T0
    #0 0x68b574 in tftp_print XYZ/tcpdump/./print-tftp.c:147:8
    #1 0x691456 in udp_print XYZ/tcpdump/./print-udp.c:549:4
    #2 0x58f360 in ip_print_demux XYZ/tcpdump/./print-ip.c:383:3
    #3 0x592006 in ip_print XYZ/tcpdump/./print-ip.c:648:3
    #4 0x566f7c in ethertype_print XYZ/tcpdump/./print-ether.c:325:10
    #5 0x566418 in ether_print XYZ/tcpdump/./print-ether.c:229:7
    #6 0x501b12 in pretty_print_packet XYZ/tcpdump/./print.c:339:18
    #7 0x4eebeb in print_packet XYZ/tcpdump/./tcpdump.c:2492:2
    #8 0x7adaf7 in pcap_offline_read XYZ/libpcap/./savefile.c:527:4
    #9 0x6b6f4c in pcap_loop XYZ/libpcap/./pcap.c:890:8
    #10 0x4ea1db in main XYZ/tcpdump/./tcpdump.c:1996:12
    #11 0x7f8c0c3eb82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)
    #12 0x419ff8 in _start (/usr/local/sbin/tcpdump+0x419ff8)

0x61d00001f281 is located 1 bytes to the right of 2048-byte region [0x61d00001ea80,0x61d00001f280)
allocated by thread T0 here:
    #0 0x4b880c in malloc /home/development/llvm/3.9.0/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:64:3
    #1 0x7b37a7 in pcap_ng_check_header XYZ/libpcap/./sf-pcap-ng.c:861:14
    #2 0x7acfd4 in pcap_fopen_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:400:7
    #3 0x7accaf in pcap_open_offline_with_tstamp_precision XYZ/libpcap/./savefile.c:307:6
    #4 0x7f8c0c3eb82f in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x2082f)

SUMMARY: AddressSanitizer: heap-buffer-overflow XYZ/tcpdump/./print-tftp.c:147:8 in tftp_print
Shadow bytes around the buggy address:
  0x0c3a7fffbe00: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fffbe10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fffbe20: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fffbe30: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c3a7fffbe40: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c3a7fffbe50:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fffbe60: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fffbe70: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fffbe80: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fffbe90: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c3a7fffbea0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==1391==ABORTING
```

**Comment by fxlb** [comment]
This one is already fixed in the security branch (will be merged in a few weeks).

**Comment by infrastation** [comment]
The original reporter for this vulnerability is Hanno Böck. It took a long time to publish the fix so the same issue was later independently discovered by other researchers.

Commit 7bf069c fixes this and the fix is available in the release 4.9.0. Thanks to everybody for their reports, please send any future security reports to security@tcpdump.org.
